osu.Game/OsuGameBase.cs

```diff
@@ -626,15 +625,40 @@ private void onRulesetChanged(ValueChangedEvent<RulesetInfo> r)
                 return;
             }
 
-            var previouslySelectedMods = SelectedMods.Value.ToArray();
+            var previouslySelectedCommonMods = new List<Mod>(SelectedMods.Value.Count);
```

Comment:
I would hope to see this changed to something much more simple:

```diff
diff --git a/osu.Game/OsuGameBase.cs b/osu.Game/OsuGameBase.cs
index e6d6deabe9..7dfdc81fb4 100644
--- a/osu.Game/OsuGameBase.cs
+++ b/osu.Game/OsuGameBase.cs
@@ -625,40 +625,31 @@ private void onRulesetChanged(ValueChangedEvent<RulesetInfo> r)
                 return;
             }
 
-            var previouslySelectedCommonMods = new List<Mod>(SelectedMods.Value.Count);
-            var convertedCommonMods = new List<Mod>(SelectedMods.Value.Count);
+            var newMods = new List<Mod>();
 
-            foreach (var mod in SelectedMods.Value)
+            // When switching rulesets, attempt to keep mods selected where they exist across both rulesets.
+            // We use the acronym to match.
+            foreach (var m in SelectedMods.Value)
             {
-                var convertedMod = instance.CreateModFromAcronym(mod.Acronym);
+                var newMod = instance.CreateModFromAcronym(m.Acronym);
 
-                if (convertedMod == null)
+                if (newMod == null)
                     continue;
 
-                previouslySelectedCommonMods.Add(mod);
-
-                convertedMod.CopyCommonSettings(mod);
-                convertedCommonMods.Add(convertedMod);
+                newMod.CopyCommonSettings(m);
+                newMods.Add(newMod);
             }
 
-            if (!ModUtils.CheckValidForGameplay(convertedCommonMods, out var invalid))
+            if (!ModUtils.CheckValidForGameplay(newMods, out var invalid))
             {
-                invalid.ForEach(mod =>
-                {
-                    int index = convertedCommonMods.IndexOf(mod);
-                    convertedCommonMods.RemoveAt(index);
-                    previouslySelectedCommonMods.RemoveAt(index);
-                });
+                // ensure we always have a valid set of mods.
+                newMods = newMods.Except(invalid).ToList();
             }
 
-            if (!SelectedMods.Disabled)
-                // Select common mods to play the deselect samples for other mods
-                SelectedMods.Value = previouslySelectedCommonMods;
-
             AvailableMods.Value = dict;
 
             if (!SelectedMods.Disabled)
-                SelectedMods.Value = convertedCommonMods;
+                SelectedMods.Value = newMods;
 
             void revertRulesetChange() => Ruleset.Value = r.OldValue?.Available == true ? r.OldValue : RulesetStore.AvailableRulesets.First();
         }

```

Reply:
This can be simplified to this
```cs
            AvailableMods.Value = dict;

            if (SelectedMods.Disabled)
                return;

            var convertedMods = SelectedMods.Value.Select(mod =>
            {
                var newMod = instance.CreateModFromAcronym(mod.Acronym);
                newMod?.CopyCommonSettings(mod);
                return newMod;
            }).Where(newMod => newMod != null).ToList();

            if (!ModUtils.CheckValidForGameplay(convertedMods, out var invalid))
                invalid.ForEach(newMod => convertedMods.Remove(newMod));

            SelectedMods.Value = convertedMods;
``` (didn't test)
with the downside of not any deselect/select samples playing, but maybe that's a good thing?

Reply:
I don't get the samples thing. Are you trying to make them play when no mods have actually changed? Or trying to inhibit when a shared mod is applied in both rulesets?

Reply:
I especially want to not play anything when the mods didn't change, only play the deselect for mods that weren't found in the ruleset we change to

Reply:
Put it this way:

You shouldn't be trying to do ANYTHING related to sample playback in this method. So remove all that and if it's an issue, we'll fix it somewhere else.

Reply:
Worth pointing out that after `AvailableMods.Value = dict;` happens no more sound will be played.

Reply:
New version seems to work well enough.